tests/e2e/runtime.go

```diff
@@ -0,0 +1,465 @@
+			return err
+		}
+	}
+	for _, n := range sc.Net.ComputeWorkers() {
+		if err := checkSynced(&n.Node); err != nil {
+			return err
+		}
+	}
+	for _, n := range sc.Net.Clients() {
+		if err := checkSynced(&n.Node); err != nil {
+			return err
+		}
+	}
+
+	sc.Logger.Info("nodes synced")
+	return nil
+}
+
+func (sc *runtimeImpl) Run(childEnv *env.Env) error {
+	// Start the test network.
+	if err := sc.Net.Start(); err != nil {
+		return err
+	}
+
+	// Wait for all nodes to sync.
+	if err := sc.waitNodesSynced(); err != nil {
+		return err
+	}
+
+	// Connect to the client node.
+	clients := sc.Net.Clients()
+	if len(clients) == 0 {
+		return fmt.Errorf("client initialization failed")
+	}
+
+	conn, err := grpc.Dial(clients[0].SocketPath(),
+		grpc.WithDefaultCallOptions(grpc.ForceCodec(&cmnGrpc.CBORCodec{})),
+		grpc.WithInsecure(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}))
+	if err != nil {
+		return err
+	}
+	rtc := client.New(conn, runtimeID)
+
+	// Run the given tests for this runtime.
+	for _, test := range sc.RunTest {
+		testName := runtime.FuncForPC(reflect.ValueOf(test).Pointer()).Name()
+
+		sc.Logger.Info("running test", "test", testName)
+		if testErr := test(sc.Net, rtc); testErr != nil {
+			sc.Logger.Error("test failed",
+				"test", testName,
+				"err", testErr,
+			)
+			return testErr
+		}
+		sc.Logger.Info("test passed", "test", testName)
+	}
+
+	return sc.Net.CheckLogWatchers()
+}
+
+// Ugh, this is really annoying :/
+//
+// cannot use sw.signer.Public() (type "github.com/oasisprotocol/oasis-core/go/common/crypto/signature".PublicKey) as type "github.com/oasisprotocol/oasis-sdk/client-sdk/go/crypto/signature".PublicKey in return argument:
+//	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature".PublicKey does not implement "github.com/oasisprotocol/oasis-sdk/client-sdk/go/crypto/signature".PublicKey (wrong type for Equal method)
+//		have Equal("github.com/oasisprotocol/oasis-core/go/common/crypto/signature".PublicKey) bool
+//		want Equal("github.com/oasisprotocol/oasis-sdk/client-sdk/go/crypto/signature".PublicKey) bool
+type pkWrapper struct {
+	pk coreSig.PublicKey
+}
+
+func newPublicKeyWrapper(pk coreSig.PublicKey) *pkWrapper {
+	return &pkWrapper{pk: pk}
+}
+
+func (pk *pkWrapper) String() string {
+	return pk.pk.String()
+}
+
+func (pk *pkWrapper) Equal(other signature.PublicKey) bool {
+	// Can't compare bytes, because signature.PublicKey can't be sliced...
+	return pk.String() == other.String()
+}
+
+func (pk *pkWrapper) Verify(context, message, signature []byte) bool {
+	return pk.pk.Verify(coreSig.Context(context), message, signature)
+}
+
+// Also really annoying:
+//
+// cannot use signer (type "github.com/oasisprotocol/oasis-core/go/common/crypto/signature".Signer) as type "github.com/oasisprotocol/oasis-sdk/client-sdk/go/crypto/signature".Signer in argument to kvInsert:
```

Comment:
https://github.com/oasisprotocol/oasis-sdk/blob/fdda3859da1fde8b8b9aab28ec9d3e55315f62c9/client-sdk/go/crypto/signature/ed25519/signer.go#L30 we have this though

Reply:
Oh, nice!  Don't know how I could have missed that :man_facepalming:  I'll fix this, thanks :)